### PR TITLE
backend: (riscv) allocate registers in function bodies only

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

**Review by nazavode — APPROVED** [review]

File: xdsl/backend/riscv/register_allocation.py [hunk 1]
```diff
@@ -1,25 +1,30 @@
...
 
-from xdsl.dialects import riscv_scf
+from xdsl.dialects import riscv_func, riscv_scf
 from xdsl.dialects.builtin import ModuleOp
 from xdsl.dialects.riscv import FloatRegisterType, IntRegisterType, RISCVOp
 from xdsl.ir import SSAValue
 
 
-class RegisterAllocator(ABC):
+class RegisterAllocator(abc.ABC):
```
Comment (AntonLydike) [inline xdsl/backend/riscv/register_allocation.py:9]: Wait, isn't this something that we didn't want to do? I remember this being a nono with `dataclasses.dataclass`?

**Review by AntonLydike — APPROVED** [review]

File: xdsl/backend/riscv/register_allocation.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/backend/riscv/register_allocation.py:9]: I have no idea. Maybe there's a ruff rule for this? I would just not think about it unless there's a reason to do it one way or another

Comment (kingiler) [reply xdsl/backend/riscv/register_allocation.py]: Excuse me, what's the issue here?

Comment (superlopuh) [reply xdsl/backend/riscv/register_allocation.py]: I think `from abc import ABC` vs `import abc`

Comment (kingiler) [reply xdsl/backend/riscv/register_allocation.py]: I see. I don't think there is such rules and it is perfectly legit usage?

**Pull request merged**